code/network/lib/Loot/Network/Class.hs

```diff
@@ -0,0 +1,176 @@
+
+
+-- This is a module containing almost all the documentation on the
+-- networking for now.
+
+{- |
+
+Networking prototype. This module aims to provide a networking API and
+its ZMQ instantiation.
+
+Requirements
+============
+
+Two basic interfaces -- client and server. Two usecases: full nodes
+and client nodes. Two APIs: client nodes use client API, full nodes
+use both. Server-only node doesn't make a lot of sense, since full
+node should talk to other full nodes (get updates etc), but it is
+possible to launch it.
+
+Client API should:
+* Receive updates from server nodes (subscriptions, (k,v) style).
+* Receive heartbeats from servers.
+* Asynchronously talk with server nodes (any node or a particular one).
+* Maintain list of peers (disconnect/connect) so we can later extend clients
+  with a discovery mechanism.
+* Talk with server nodes in separate "client" (worker) threads.
+Not required:
+* Sending heartbeats to server (heartbeating server -> client is sufficient,
+  all full nodes are servers).
+
+Server API should:
+* Respond to users' requests asynchronously.
+* Responses should be done in separate "listener" threads, each typically
+  blocking until some request comes and then replying in a free manner until
+  it decides to block again.
+* Publish any (k,v) data, such as transactions, headers or blocks.
+* Send heartbeats to clients.
+Not required:
+* Maintaining list of clients and replying first (without request), that's
+  what "publishing" is for.
+
+-}
+
+module Loot.Network.Class
+    ( Content
+    , Subscription (..)
+    , MsgType (..)
+    , BiTQueue (..)
+
+    , CliRecvMsg(..)
+    , ClientId
+    , ClientEnv
+    , UpdatePeersReq (..)
+    , uprAdd
+    , uprDel
+    , NetworkingCli (..)
+
+    , ListenerId
+    , ListenerEnv
+    , ServSendMsg (..)
+    , NetworkingServ (..)
+    ) where
+
+import Control.Concurrent.STM.TQueue (TQueue)
+import Control.Lens (makeLenses)
+import Data.Default (Default (..))
+
+----------------------------------------------------------------------------
+-- Common
+----------------------------------------------------------------------------
+
+-- | We send and receive lists of strict bytestring. ZMQ
+-- implementation operates in frames, so user can use this extra
+-- functionlaity (for example, attach message numbers in the fisrt
+-- frame and data in the second one). If it is not needed, your
+-- protocol design can state that one or zero BSs are sent.
+type Content = [ByteString]
+
+-- | Bytestring describing the type of the subscription. The key.
+newtype Subscription = Subscription { unSubscription :: ByteString } deriving (Eq,Ord,Show,IsString)
+
+-- | Message type is characterized as a bytestring.
+newtype MsgType = MsgType { unMsgType :: ByteString } deriving (Eq,Ord,Show,IsString)
+
+-- | Two-ended queue, first parameter is messages user receives,
+-- second -- messages user sends.
+data BiTQueue r s = BiTQueue
+    { bReceiveQ :: TQueue r
```

Comment:
I thought about it for a while and I won't replace it. The reason is that both brokers and listeners/clients are supposed to be asynchronous proxies which can't normally block until the queue will get free. We assume that router can read and write to the queues with much faster rate than the data is actually produced. By design, the broker can't block on writing to the receive queue, so it can't be `TBQueue` (maybe we can just drop the data if the queue is full?). I expect that related listener/client will be running all the actual work in separate threads and the main thread only will be connected to the queue, working as proxy. So in case a lot of data comes, it (client/listener proxy) will take responsibility to discard it.